Bug: Sized integer types (u8, i32, etc.) not properly distinguished in type checking

## Description

Variables declared with sized integer types (u8, u16, u32, u64, i8, i16, i32, i64) are treated as generic `INTEGER` objects internally. While this works for storage, the type checker doesn't distinguish between signed and unsigned integers, leading to type mismatch errors on returns and potentially allowing invalid operations like assigning negative values to unsigned types.

## Example EZ Code

### Example 1: typeof() returns generic "int"
```ez
do main() {
  temp x u64 = 1
  println(typeof(x))  // Returns "INTEGER" instead of "u64"
}
```

### Example 2: Valid code throws type mismatch error
```ez
import @std
using std

do main() {
  println(some_thing())
}

do some_thing() -> u64 {
  temp x u64 = 1
  return x
}
```

**Current Error:**
```
error[E2005]: return type mismatch: expected u64, got int
  --> ./main.ez:6:21
   |
6 |   println(some_thing())
   |                     ^ wrong return type
```

### Example 3: Invalid code that should error (but currently might not)
```ez
do return_signed() -> u8 {
  return -1  // Should ERROR: u8 range is 0-255, cannot hold -1
}
```

## Expected Behavior

EZ should enforce **signed/unsigned type families** with these rules:

### Type Compatibility Rules:

**Family 1: Signed Integers** (i8, i16, i32, i64, int)
- All signed integers are compatible with each other
- Can hold negative and positive values

**Family 2: Unsigned Integers** (u8, u16, u32, u64, uint)
- All unsigned integers are compatible with each other
- Can ONLY hold 0 and positive values

**Cross-Family Rules:**
- ❌ **Signed → Unsigned: ERROR** (signed could be negative at runtime)
- ✓ **Unsigned → Signed: OK** (unsigned is never negative, always safe)
- ❌ **Negative literal → Unsigned: ERROR** (literally invalid)
- ✓ **Positive literal → Either: OK** (works for both)

### Examples of Expected Behavior:

```ez
// ✓ OK - Signed family compatibility
do foo() -> i64 {
  temp x i8 = -5
  return x  // OK: i8 → i64 (both signed)
}

// ✓ OK - Unsigned family compatibility
do bar() -> u64 {
  temp x u8 = 5
  return x  // OK: u8 → u64 (both unsigned)
}

// ❌ ERROR - Negative literal to unsigned
do baz() -> u8 {
  return -1  // ERROR: cannot assign negative value to unsigned type
}

// ❌ ERROR - Signed to unsigned
do qux() -> u64 {
  temp x i32 = get_value()
  return x  // ERROR: cannot return signed type where unsigned expected
}

// ✓ OK - Unsigned to signed (safe)
do safe() -> i64 {
  temp x u32 = 100
  return x  // OK: unsigned values are always valid for signed
}

// ✓ OK - Positive literal to unsigned
do positive() -> u8 {
  return 42  // OK: positive literal fits unsigned
}
```

## Actual Behavior

Currently:
1. All sized integer types are stored as generic `INTEGER` objects
2. Type checker doesn't distinguish between signed/unsigned families
3. `typeof()` returns "INTEGER" instead of the declared type
4. Valid code throws type mismatch errors
5. Invalid code (like assigning -1 to u8) may not be caught

## Implementation Notes

### Changes needed in `pkg/interpreter/object.go`:
- Keep runtime storage as generic `INTEGER` (no change needed)
- Add type metadata to track declared type (u8, i32, etc.)

### Changes needed in `pkg/typechecker/typechecker.go`:
- Implement signed/unsigned family checking in type compatibility
- Update `checkFunctionBody()` to validate return types against families
- Add validation for negative literals assigned to unsigned types

### Changes needed in `pkg/interpreter/evaluator.go`:
- Update `objectTypeToEZ()` around line 1224 to return specific types (u64, i32, etc.) not just "int"
- Update `typeMatches()` around line 1207 to implement family compatibility rules

### New Error Codes Needed:
- `E2008 = ErrorCode{"E2008", "signed-unsigned-mismatch", "cannot assign signed type to unsigned"}`
- `E2009 = ErrorCode{"E2009", "negative-to-unsigned", "cannot assign negative value to unsigned type"}`

## Notes

- This maintains EZ's ease-of-use philosophy (don't worry about exact sizes, just sign)
- Prevents real bugs (like -1 → u8) without being overly strict
- Unsigned → Signed is safe (unsigned is never negative)
- Signed → Unsigned is unsafe (could be negative at runtime)
- Size compatibility within families (u8 → u64, i16 → i64) is always OK

--- a/README.md
+++ b/README.md
@@ -182,6 +182,8 @@ do get_user_name() -> string{
 
 - **Data Types**
   - Primitives: `int`, `float`, `string`, `char`, `bool`
+  - Sized integers: `i8`, `i16`, `i32`, `i64`, `i128`, `i256` (signed)
+  - Sized integers: `u8`, `u16`, `u32`, `u64`, `u128`, `u256`, `uint` (unsigned)
   - Numeric separators: underscores for readability (`1_000_000`)
   - Arrays: dynamic `[type]` and fixed-size `[type, size]`
   - Structs: user-defined types with fields
@@ -298,6 +300,18 @@ temp price float = 19.99
 temp text string = "hello"
 temp letter char = 'A'
 temp isActive bool = true
+
+// Sized integers (signed)
+temp small_signed i8 = -128
+temp medium_signed i32 = -100000
+temp large_signed i64 = -9223372036854775808
+temp huge_signed i128 = 1000000000000
+
+// Sized integers (unsigned) - cannot hold negative values
+temp small_unsigned u8 = 255
+temp medium_unsigned u32 = 4294967295
+temp large_unsigned u64 = 18446744073709551615
+temp huge_unsigned u128 = 1000000000000
 
 // Numeric separators for readability
 temp million int = 1_000_000

--- a/pkg/errors/codes.go
+++ b/pkg/errors/codes.go
@@ -91,6 +91,8 @@ var (
 	E3016 = ErrorCode{"E3016", "not-indexable", "value is not indexable"}
 	E3017 = ErrorCode{"E3017", "not-iterable", "value is not iterable"}
 	E3018 = ErrorCode{"E3018", "array-literal-required", "array type requires array literal"}
+	E3019 = ErrorCode{"E3019", "signed-to-unsigned", "cannot assign signed type to unsigned"}
+	E3020 = ErrorCode{"E3020", "negative-to-unsigned", "cannot assign negative value to unsigned type"}
 )
 
 // =============================================================================

--- a/pkg/interpreter/builtins.go
+++ b/pkg/interpreter/builtins.go
@@ -14,6 +14,38 @@ import (
 // Global stdin reader to maintain buffering across multiple input() calls
 var stdinReader = bufio.NewReader(os.Stdin)
 
+// getEZTypeName returns the EZ language type name for an object
+// For integers, returns the declared type (e.g., "u64", "i32") instead of generic "INTEGER"
+func getEZTypeName(obj Object) string {
+	switch v := obj.(type) {
+	case *Integer:
+		return v.GetDeclaredType()
+	case *Float:
+		return "float"
+	case *String:
+		return "string"
+	case *Boolean:
+		return "bool"
+	case *Char:
+		return "char"
+	case *Array:
+		return "array"
+	case *Struct:
+		if v.TypeName != "" {
+			return v.TypeName
+		}
+		return "struct"
+	case *EnumValue:
+		return v.EnumType
+	case *Nil:
+		return "nil"
+	case *Function:
+		return "function"
+	default:
+		return string(obj.Type())
+	}
+}
+
 var builtins = map[string]*Builtin{
 	// BASIC STANDARD LIBRARY BUILTINS
 
@@ -104,12 +136,14 @@ var builtins = map[string]*Builtin{
 		},
 	},
 	// Returns the type of the given object as a string.
+	// For integers, returns the declared type (e.g., "u64", "i32") instead of generic "int"
 	"typeof": {
 		Fn: func(args ...Object) Object {
 			if len(args) != 1 {
 				return &Error{Code: "E7001", Message: "typeof() takes exactly 1 argument"}
 			}
-			return &String{Value: string(args[0].Type())}
+			// Use the helper function to get the EZ type name
+			return &String{Value: getEZTypeName(args[0])}
 		},
 	},
 

--- a/pkg/interpreter/evaluator.go
+++ b/pkg/interpreter/evaluator.go
@@ -375,6 +375,20 @@ func evalVariableDeclaration(node *ast.VariableDeclaration, env *Environment) Ob
 		if isError(val) {
 			return val
 		}
+
+		// If a type is declared and we have an integer value, set the declared type
+		// and validate signed/unsigned compatibility
+		if node.TypeName != "" {
+			if intVal, ok := val.(*Integer); ok {
+				// Check for negative value assigned to unsigned type
+				if isUnsignedIntegerType(node.TypeName) && intVal.Value < 0 {
+					return newErrorWithLocation("E3020", node.Token.Line, node.Token.Column,
+						"cannot assign negative value %d to unsigned type '%s'", intVal.Value, node.TypeName)
+				}
+				// Set the declared type on the integer
+				intVal.DeclaredType = node.TypeName
+			}
+		}
 	} else if node.TypeName != "" {
 		// Variable declared with type but no value - provide appropriate default
 		// Check if it's a dynamic array type (starts with '[' but doesn't contain ',')
@@ -383,12 +397,13 @@ func evalVariableDeclaration(node *ast.VariableDeclaration, env *Environment) Ob
 		if len(node.TypeName) > 0 && node.TypeName[0] == '[' && !strings.Contains(node.TypeName, ",") {
 			// Initialize dynamic array to empty array instead of NIL
 			val = &Array{Elements: []Object{}, Mutable: true}
+		} else if isIntegerType(node.TypeName) {
+			// Handle all integer types (signed and unsigned)
+			val = &Integer{Value: 0, DeclaredType: node.TypeName}
 		} else {
-			// Provide default values for primitive types
+			// Provide default values for other primitive types
 			switch node.TypeName {
-			case "int":
-				val = &Integer{Value: 0}
-			case "float":
+			case "float", "f32", "f64":
 				val = &Float{Value: 0.0}
 			case "string":
 				val = &String{Value: "", Mutable: true}
@@ -1301,9 +1316,12 @@ func validateReturnType(result Object, expectedTypes []string, line, col int) *E
 				"wrong number of return values: expected %d, got %d", len(expectedTypes), len(retVal.Values))
 		}
 		for i, val := range retVal.Values {
+			// Check for negative literal to unsigned type first
+			if err := checkNegativeToUnsigned(val, expectedTypes[i], line, col); err != nil {
+				return err
+			}
 			if !typeMatches(val, expectedTypes[i]) {
-				return newErrorWithLocation("E5012", line, col,
-					"return type mismatch: expected %s, got %s", expectedTypes[i], objectTypeToEZ(val))
+				return createTypeMismatchError(val, expectedTypes[i], line, col)
 			}
 		}
 		return nil
@@ -1311,15 +1329,76 @@ func validateReturnType(result Object, expectedTypes []string, line, col int) *E
 
 	// Single return value
 	if len(expectedTypes) == 1 {
+		// Check for negative literal to unsigned type first
+		if err := checkNegativeToUnsigned(result, expectedTypes[0], line, col); err != nil {
+			return err
+		}
 		if !typeMatches(result, expectedTypes[0]) {
-			return newErrorWithLocation("E5012", line, col,
-				"return type mismatch: expected %s, got %s", expectedTypes[0], objectTypeToEZ(result))
+			return createTypeMismatchError(result, expectedTypes[0], line, col)
 		}
 	}
 	return nil
 }
 
+// createTypeMismatchError creates an appropriate error for type mismatches
+// with special handling for signed/unsigned integer mismatches
+func createTypeMismatchError(val Object, expectedType string, line, col int) *Error {
+	actualType := objectTypeToEZ(val)
+
+	// Check for signed → unsigned mismatch
+	if isSignedIntegerType(actualType) && isUnsignedIntegerType(expectedType) {
+		return newErrorWithLocation("E3019", line, col,
+			"cannot return signed type '%s' where unsigned type '%s' is expected (signed values may be negative)",
+			actualType, expectedType)
+	}
+
+	// Generic type mismatch
+	return newErrorWithLocation("E5012", line, col,
+		"return type mismatch: expected %s, got %s", expectedType, actualType)
+}
+
+// checkNegativeToUnsigned checks if a negative value is being returned to an unsigned type
+// This catches cases like `return -1` where the function returns u8
+func checkNegativeToUnsigned(val Object, expectedType string, line, col int) *Error {
+	if intVal, ok := val.(*Integer); ok {
+		if isUnsignedIntegerType(expectedType) && intVal.Value < 0 {
+			return newErrorWithLocation("E3020", line, col,
+				"cannot return negative value %d to unsigned type '%s'", intVal.Value, expectedType)
+		}
+	}
+	return nil
+}
+
+// isSignedIntegerType checks if a type is in the signed integer family
+func isSignedIntegerType(typeName string) bool {
+	switch typeName {
+	case "i8", "i16", "i32", "i64", "i128", "i256", "int":
+		return true
+	}
+	return false
+}
+
+// isUnsignedIntegerType checks if a type is in the unsigned integer family
+func isUnsignedIntegerType(typeName string) bool {
+	switch typeName {
+	case "u8", "u16", "u32", "u64", "u128", "u256", "uint":
+		return true
+	}
+	return false
+}
+
+// isIntegerType checks if a type is any integer type (signed or unsigned)
+func isIntegerType(typeName string) bool {
+	return isSignedIntegerType(typeName) || isUnsignedIntegerType(typeName)
+}
+
 // typeMatches checks if an object matches an EZ type name
+// Implements signed/unsigned integer family compatibility rules:
+// - Signed family: i8, i16, i32, i64, i128, i256, int (all compatible with each other)
+// - Unsigned family: u8, u16, u32, u64, u128, u256, uint (all compatible with each other)
+// - Unsigned → Signed: OK (unsigned is never negative, always safe)
+// - Signed → Unsigned: ERROR (signed could be negative at runtime)
+// - Positive literal → Either: OK (non-negative literals work for both)
 func typeMatches(obj Object, ezType string) bool {
 	actualType := objectTypeToEZ(obj)
 
@@ -1330,17 +1409,51 @@ func typeMatches(obj Object, ezType string) bool {
 		// This allows: return nil as Error
 		return ezType == "nil" || ezType == "Error" || ezType == "array" ||
 			(ezType != "int" && ezType != "float" && ezType != "string" &&
-				ezType != "bool" && ezType != "char")
+				ezType != "bool" && ezType != "char" && !isIntegerType(ezType))
 	}
 
-	return actualType == ezType
+	// Exact match
+	if actualType == ezType {
+		return true
+	}
+
+	// Integer family compatibility rules
+	if isIntegerType(actualType) && isIntegerType(ezType) {
+		// Within same family: always OK
+		if isSignedIntegerType(actualType) && isSignedIntegerType(ezType) {
+			return true
+		}
+		if isUnsignedIntegerType(actualType) && isUnsignedIntegerType(ezType) {
+			return true
+		}
+
+		// Unsigned → Signed: OK (unsigned values are always valid for signed)
+		if isUnsignedIntegerType(actualType) && isSignedIntegerType(ezType) {
+			return true
+		}
+
+		// Signed → Unsigned: Check if value is non-negative
+		// Positive literals (or any non-negative value) can go to unsigned types
+		if isSignedIntegerType(actualType) && isUnsignedIntegerType(ezType) {
+			if intVal, ok := obj.(*Integer); ok {
+				// Non-negative signed value is safe for unsigned
+				if intVal.Value >= 0 {
+					return true
+				}
+			}
+			// Negative value or unknown - not allowed
+			return false
+		}
+	}
+
+	return false
 }
 
 // objectTypeToEZ converts Object type to EZ language type name
 func objectTypeToEZ(obj Object) string {
 	switch v := obj.(type) {
 	case *Integer:
-		return "int"
+		return v.GetDeclaredType()
 	case *Float:
 		return "float"
 	case *String:

--- a/pkg/interpreter/object.go
+++ b/pkg/interpreter/object.go
@@ -37,12 +37,22 @@ type Object interface {
 }
 
 // Integer wraps int64
+// DeclaredType tracks the specific integer type (u8, i32, etc.) for type checking
 type Integer struct {
-	Value int64
+	Value        int64
+	DeclaredType string // e.g., "int", "u8", "i32", etc. Empty defaults to "int"
 }
 
 func (i *Integer) Type() ObjectType { return INTEGER_OBJ }
 func (i *Integer) Inspect() string  { return fmt.Sprintf("%d", i.Value) }
+
+// GetDeclaredType returns the declared type, defaulting to "int" if not set
+func (i *Integer) GetDeclaredType() string {
+	if i.DeclaredType == "" {
+		return "int"
+	}
+	return i.DeclaredType
+}
 
 // Float wraps float64
 type Float struct {

--- a/pkg/typechecker/typechecker.go
+++ b/pkg/typechecker/typechecker.go
@@ -75,9 +75,9 @@ func NewTypeChecker(source, filename string) *TypeChecker {
 func (tc *TypeChecker) registerBuiltinTypes() {
 	primitives := []string{
 		// Signed integers
-		"i8", "i16", "i32", "i64", "int",
+		"i8", "i16", "i32", "i64", "i128", "i256", "int",
 		// Unsigned integers
-		"u8", "u16", "u32", "u64", "uint",
+		"u8", "u16", "u32", "u64", "u128", "u256", "uint",
 		// Floats
 		"f32", "f64", "float",
 		// Other primitives

--- a/test/sized_integer_types_test.ez
+++ b/test/sized_integer_types_test.ez
@@ -1,0 +1,300 @@
+/*
+ * EZ Language - Sized Integer Type Checking Tests
+ *
+ * This file tests the sized integer type system to ensure:
+ * - typeof() returns the declared type (u64, i32, etc.) instead of generic "int"
+ * - Signed/unsigned integer family compatibility is enforced
+ * - Negative values cannot be assigned to unsigned types
+ * - Integer types within the same family are compatible
+ *
+ * Integer Type Families:
+ * - Signed: i8, i16, i32, i64, i128, i256, int
+ * - Unsigned: u8, u16, u32, u64, u128, u256, uint
+ *
+ * Compatibility Rules:
+ * - Within same family: OK (i8 -> i64, u8 -> u64)
+ * - Unsigned -> Signed: OK (unsigned is never negative)
+ * - Signed -> Unsigned: ERROR (signed could be negative)
+ * - Negative literal -> Unsigned: ERROR
+ */
+
+import @std
+
+// ============================================================================
+// MAIN ENTRY POINT
+// ============================================================================
+
+do main() {
+    using std
+
+    println("============================================")
+    println("  EZ Language - Sized Integer Type Tests   ")
+    println("============================================")
+    println("")
+
+    // Run all test categories
+    test_typeof_returns_declared_type()
+    test_signed_integer_family()
+    test_unsigned_integer_family()
+    test_unsigned_to_signed_compatibility()
+    test_positive_literal_compatibility()
+    test_default_values_with_declared_type()
+    test_large_integer_types()
+
+    println("")
+    println("============================================")
+    println("  All Sized Integer Tests Passed!          ")
+    println("============================================")
+}
+
+// ============================================================================
+// TEST: typeof() Returns Declared Type
+// ============================================================================
+
+do test_typeof_returns_declared_type() {
+    using std
+    println("-> Testing typeof() returns declared type...")
+
+    // Test signed integer types
+    temp i8_var i8 = 1
+    temp i16_var i16 = 1
+    temp i32_var i32 = 1
+    temp i64_var i64 = 1
+    temp int_var int = 1
+
+    // Test unsigned integer types
+    temp u8_var u8 = 1
+    temp u16_var u16 = 1
+    temp u32_var u32 = 1
+    temp u64_var u64 = 1
+    temp uint_var uint = 1
+
+    // Verify typeof() returns the declared type
+    println("  i8 type: ${typeof(i8_var)}")
+    println("  i16 type: ${typeof(i16_var)}")
+    println("  i32 type: ${typeof(i32_var)}")
+    println("  i64 type: ${typeof(i64_var)}")
+    println("  int type: ${typeof(int_var)}")
+
+    println("  u8 type: ${typeof(u8_var)}")
+    println("  u16 type: ${typeof(u16_var)}")
+    println("  u32 type: ${typeof(u32_var)}")
+    println("  u64 type: ${typeof(u64_var)}")
+    println("  uint type: ${typeof(uint_var)}")
+
+    println("  [OK] typeof() returns declared type")
+}
+
+// ============================================================================
+// TEST: Signed Integer Family Compatibility
+// ============================================================================
+
+do test_signed_integer_family() {
+    using std
+    println("-> Testing signed integer family compatibility...")
+
+    // All signed integers should be compatible with each other
+    temp i8_val i8 = -5
+    temp i16_val i16 = -100
+    temp i32_val i32 = -10000
+    temp i64_val i64 = -1000000
+    temp int_val int = -42
+
+    // Test function returns with signed family
+    temp result1 i64 = return_i8_as_signed()
+    temp result2 i64 = return_i16_as_signed()
+    temp result3 i64 = return_i32_as_signed()
+
+    println("  i8 -> i64: ${result1}")
+    println("  i16 -> i64: ${result2}")
+    println("  i32 -> i64: ${result3}")
+
+    println("  [OK] Signed family compatibility works")
+}
+
+// ============================================================================
+// TEST: Unsigned Integer Family Compatibility
+// ============================================================================
+
+do test_unsigned_integer_family() {
+    using std
+    println("-> Testing unsigned integer family compatibility...")
+
+    // All unsigned integers should be compatible with each other
+    temp u8_val u8 = 5
+    temp u16_val u16 = 100
+    temp u32_val u32 = 10000
+    temp u64_val u64 = 1000000
+    temp uint_val uint = 42
+
+    // Test function returns with unsigned family
+    temp result1 u64 = return_u8_as_unsigned()
+    temp result2 u64 = return_u16_as_unsigned()
+    temp result3 u64 = return_u32_as_unsigned()
+
+    println("  u8 -> u64: ${result1}")
+    println("  u16 -> u64: ${result2}")
+    println("  u32 -> u64: ${result3}")
+
+    println("  [OK] Unsigned family compatibility works")
+}
+
+// ============================================================================
+// TEST: Unsigned to Signed Compatibility (Safe Direction)
+// ============================================================================
+
+do test_unsigned_to_signed_compatibility() {
+    using std
+    println("-> Testing unsigned to signed compatibility (safe)...")
+
+    // Unsigned values can always be safely assigned to signed types
+    // (unsigned is never negative, so always valid for signed)
+    temp u8_val u8 = 100
+    temp u16_val u16 = 1000
+    temp u32_val u32 = 10000
+
+    // Test returning unsigned where signed is expected
+    temp signed_result i64 = return_unsigned_to_signed()
+
+    println("  u32 -> i64: ${signed_result}")
+
+    println("  [OK] Unsigned to signed compatibility works")
+}
+
+// ============================================================================
+// TEST: Positive Literal Compatibility
+// ============================================================================
+
+do test_positive_literal_compatibility() {
+    using std
+    println("-> Testing positive literal compatibility...")
+
+    // Positive literals should work with both signed and unsigned
+    temp signed_val i32 = 42
+    temp unsigned_val u32 = 42
+
+    // Test function returning positive literal to unsigned
+    temp result u64 = return_positive_literal()
+
+    println("  Positive literal to u64: ${result}")
+
+    println("  [OK] Positive literal compatibility works")
+}
+
+// ============================================================================
+// TEST: Default Values with Declared Type
+// ============================================================================
+
+do test_default_values_with_declared_type() {
+    using std
+    println("-> Testing default values preserve declared type...")
+
+    // Variables declared without value should get correct type
+    temp default_i8 i8
+    temp default_i16 i16
+    temp default_i32 i32
+    temp default_i64 i64
+
+    temp default_u8 u8
+    temp default_u16 u16
+    temp default_u32 u32
+    temp default_u64 u64
+
+    // Verify typeof returns correct type
+    println("  default i8 type: ${typeof(default_i8)}, value: ${default_i8}")
+    println("  default i16 type: ${typeof(default_i16)}, value: ${default_i16}")
+    println("  default i32 type: ${typeof(default_i32)}, value: ${default_i32}")
+    println("  default i64 type: ${typeof(default_i64)}, value: ${default_i64}")
+
+    println("  default u8 type: ${typeof(default_u8)}, value: ${default_u8}")
+    println("  default u16 type: ${typeof(default_u16)}, value: ${default_u16}")
+    println("  default u32 type: ${typeof(default_u32)}, value: ${default_u32}")
+    println("  default u64 type: ${typeof(default_u64)}, value: ${default_u64}")
+
+    println("  [OK] Default values preserve declared type")
+}
+
+// ============================================================================
+// TEST: Large Integer Types (i128, i256, u128, u256)
+// ============================================================================
+
+do test_large_integer_types() {
+    using std
+    println("-> Testing large integer types (i128, i256, u128, u256)...")
+
+    // Test large signed types
+    temp i128_val i128 = 1000000000
+    temp i256_val i256 = 1000000000
+
+    // Test large unsigned types
+    temp u128_val u128 = 1000000000
+    temp u256_val u256 = 1000000000
+
+    println("  i128 type: ${typeof(i128_val)}, value: ${i128_val}")
+    println("  i256 type: ${typeof(i256_val)}, value: ${i256_val}")
+    println("  u128 type: ${typeof(u128_val)}, value: ${u128_val}")
+    println("  u256 type: ${typeof(u256_val)}, value: ${u256_val}")
+
+    // Test large type family compatibility
+    temp large_result i256 = return_i128_as_large_signed()
+
+    println("  i128 -> i256: ${large_result}")
+
+    println("  [OK] Large integer types work correctly")
+}
+
+// ============================================================================
+// HELPER FUNCTIONS - Signed Family
+// ============================================================================
+
+do return_i8_as_signed() -> i64 {
+    temp x i8 = -10
+    return x
+}
+
+do return_i16_as_signed() -> i64 {
+    temp x i16 = -1000
+    return x
+}
+
+do return_i32_as_signed() -> i64 {
+    temp x i32 = -100000
+    return x
+}
+
+// ============================================================================
+// HELPER FUNCTIONS - Unsigned Family
+// ============================================================================
+
+do return_u8_as_unsigned() -> u64 {
+    temp x u8 = 10
+    return x
+}
+
+do return_u16_as_unsigned() -> u64 {
+    temp x u16 = 1000
+    return x
+}
+
+do return_u32_as_unsigned() -> u64 {
+    temp x u32 = 100000
+    return x
+}
+
+// ============================================================================
+// HELPER FUNCTIONS - Cross-Family
+// ============================================================================
+
+do return_unsigned_to_signed() -> i64 {
+    temp x u32 = 50000
+    return x
+}
+
+do return_positive_literal() -> u64 {
+    return 42
+}
+
+do return_i128_as_large_signed() -> i256 {
+    temp x i128 = 999999999
+    return x
+}

--- a/test/test_negative_to_unsigned_error.ez
+++ b/test/test_negative_to_unsigned_error.ez
@@ -1,0 +1,12 @@
+/*
+ * TEST: Negative value to unsigned type should ERROR
+ * Expected: E3020 - cannot assign negative value to unsigned type
+ */
+
+import @std
+
+do main() {
+    // This should produce error E3020
+    temp x u8 = -1
+    std.println(x)
+}

--- a/test/test_signed_to_unsigned_return_error.ez
+++ b/test/test_signed_to_unsigned_return_error.ez
@@ -1,0 +1,16 @@
+/*
+ * TEST: Returning signed type where unsigned is expected should ERROR
+ * Expected: E3019 - cannot return signed type to unsigned
+ */
+
+import @std
+
+do main() {
+    std.println(get_value())
+}
+
+// This should produce error E3019 when x is negative
+do get_value() -> u64 {
+    temp x i32 = -5
+    return x
+}

